src/java.compiler/share/classes/javax/tools/JavaFileManager.java

```diff
@@ -338,6 +340,49 @@ JavaFileObject getJavaFileForOutput(Location location,
                                         FileObject sibling)
         throws IOException;
 
+    /**
+     * Returns a {@linkplain JavaFileObject file object} for output
+     * representing the specified class of the specified kind in the
+     * given package-oriented location.
+     *
+     * <p>The provided {@code originatingFiles} represent files that
+     * were in, an unspecified way, used to create the content of
+     * the file created by this method. See {@code originatingElements}
+     * in {@link Filer#createSourceFile}. {@link Elements#getFileObjectOf}
+     * may be used to convert the {@code Element}s to {@code FileObject}s.
+     *
+     * @implSpec The default implementation calls
+     * {@link #getJavaFileForOutput(javax.tools.JavaFileManager.Location, java.lang.String, javax.tools.JavaFileObject.Kind, javax.tools.FileObject) }
+     * with the first element of the {@code originatingFiles}, if any, as a {@code sibling}.
+     *
+     * @param location a package-oriented location
+     * @param className the name of a class
+     * @param kind the kind of file, must be one of {@link
+     * JavaFileObject.Kind#SOURCE SOURCE} or {@link
+     * JavaFileObject.Kind#CLASS CLASS}
+     * @param originatingFiles the files which are contributing to this newly created file
+     * @return a file object for output
+     * @throws IllegalArgumentException if sibling is not known to
+     * this file manager, or if the location is not known to this file
+     * manager and the file manager does not support unknown
+     * locations, or if the kind is not valid, or if the location is
+     * not an output location
+     * @throws IOException if an I/O error occurred, or if {@link
+     * #close} has been called and this file manager cannot be
+     * reopened
+     * @throws IllegalStateException {@link #close} has been called
+     * and this file manager cannot be reopened
+     * @since 18
+     * @see Filer#createSourceFile
+     */
+    default JavaFileObject getJavaFileForOutputForOriginatingFiles(Location location,
+                                        String className,
+                                        Kind kind,
+                                        FileObject... originatingFiles)
```

Comment:
Perhaps explicitly state that null originatingFile or empty originatingFiles are no-ops.

Reply:
Thanks - how about the following?
https://github.com/openjdk/jdk/pull/5076/commits/7c11f10d9c2bce6a5841fcfc4580edc12692dde5

Reply:
Spec update looks fine.